### PR TITLE
xds: perform header matching on concatenated values

### PR DESCRIPTION
**Pull request by voidzcy**

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/RouteMatch.java [hunk 1]
```diff
@@ -60,7 +60,7 @@
    *
    * <p>Match is not deterministic if a runtime fraction match rule presents in this RouteMatch.
    */
-  boolean matches(String path, Map<String, Set<String>> headers) {
+  boolean matches(String path, Map<String, Iterable<String>> headers) {
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/RouteMatch.java:63]: Why not just `boolean matches(String path, Metadata headers)` without making a copy map?  
```java
for (HeaderMatcher headerMatcher : headerMatchers) {
  headerName = headerMatcher.getName();
  Iterable<String> values = headers.getAll(
      Metadata.Key.of(headerName, Metadata.ASCII_STRING_MARSHALLER));
  if (headerMatcher.matchesValue(values)) {
    return false;
  }
}
```

Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/RouteMatch.java:63]: Seems the copy map can be reused for multiple RouteMatches in pickSubchannel(), which is more efficient.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/RouteMatch.java:63]: Yes, exactly.

**Comment by dapengzhang0** [comment]
I think we need backport it.

**Pull request merged**